plutus-core/plutus-ir/src/PlutusIR/Compiler/Datatype.hs

```diff
@@ -452,9 +452,8 @@ mkConstructor opts dty d@(Datatype ann tv tvs _ constrs) index = do
 
             -- The pattern functor with a hole in it
             pf <- mkPatternFunctorBody opts ann d
-            -- ... and with the hole filled in with the datatype type, by making
-            -- a type lambda and immediately applying it
-            let unrolled = TyApp ann (PIR.mkIterTyLam [tv] pf) (getType dty)
+            -- ... and with the hole filled in with the datatype type
+            let unrolled = unveilDatatype (getType dty) d pf
```

Comment:
The only thing I'm still worried about here if the parameters in `pf` can somehow get out of sync with those of `d` (i.e. `fmap (PIR.mkTyVar ann) tvs`).

And yeah, it's still weird to reconstruct the pattern functor body.